src/device/cpu/riscv_rv32ima/tlb.c

```diff
@@ -0,0 +1,224 @@
+#include <string.h>
+
+#include "tlb.h"
+#include "../../../utils.h"
+
+typedef struct rv_tlb_entry {
+    sv32_pte_t pte;
+    uint32_t vpn;
+    unsigned asid;
+    bool valid;
+    bool global;
+} rv_tlb_entry_t;
+
+static inline size_t hash(size_t num){
+    // No hashing as of now
+    return num;
+}
+
+#define index_ktlb(tlb, index) \
+    ((tlb)->ktlb[hash((index)) % ((tlb)->ktlb_size)])
+
+#define index_mtlb(tlb, index) \
+    ((tlb)->mtlb[hash((index)) % ((tlb)->mtlb_size)])
+
+/** Caches a mapping into the TLB */
+extern void rv_tlb_add_mapping(rv_tlb_t* tlb, unsigned asid, uint32_t virt, sv32_pte_t pte, bool megapage, bool global){
+    
+    uint32_t vpn = virt >> RV_PAGESIZE;
+    
+    if(megapage){
+        uint32_t mvpn = virt >> RV_MEGAPAGESIZE;
+
+        index_mtlb(tlb, mvpn).vpn = mvpn;
+        index_mtlb(tlb, mvpn).pte = pte;
+        index_mtlb(tlb, mvpn).global = global;
+        index_mtlb(tlb, mvpn).asid = asid;
+        index_mtlb(tlb, mvpn).valid = true;    
+    }
+    else {
+        index_ktlb(tlb, vpn).vpn = vpn;
+        index_ktlb(tlb, vpn).pte = pte;
+        index_ktlb(tlb, vpn).global = global;
+        index_ktlb(tlb, vpn).asid = asid;
+        index_ktlb(tlb, vpn).valid = true;
+    }
+}
+
+/** Retrieves a cached mapping
+ * gives priority to megapage mappings
+ */
+extern bool rv_tlb_get_mapping(rv_tlb_t* tlb, unsigned asid, uint32_t virt, sv32_pte_t* pte, bool* megapage){
+    uint32_t vpn = virt >> 12;
+    uint32_t mvpn = vpn >> 10;
```

Comment:
Probably `RV_PAGESIZE` instead of the constant should be used?

Reply:
Yes, thanks.
Fixed in [d8210af](https://github.com/d-iii-s/msim/pull/36/commits/d8210afea32d8c76b88054151fed5de807a85116)